### PR TITLE
WIP ENH add SCAD penalty

### PR DESCRIPTION
**Pull request by PABannier**
Revival of old branch. The goal of this branch is to add the SCAD and BlockSCAD penalties.

### CODE REVIEW

**Comment by PABannier** [comment]
I copy-pasted the work already done by @Klopfe in the previous repo. I'm working on BlockSCAD.

@Klopfe @QB3 @mathurinm any ideas on against which solver/framework we should test SCAD (also MCP, and L0.5). The non-convex penalties are not tested at the moment.

File: skglm/penalties/block_separable.py [hunk 1]
```diff
@@ -153,3 +153,81 @@ def subdiff_distance(self, W, grad, ws):
...
+
+    def __init__(self, alpha, gamma):
+        self.alpha = alpha
+        self.gamma = gamma
+
+    def value(self, W):
+        """Compute the value of the SCAD penalty at W."""
+        n_features = W.shape[1]
+        norm_rows = np.sqrt(np.sum(W ** 2, axis=1))
+        value = np.full_like(norm_rows, ((self.gamma + 1) * self.alpha ** 2) / 2.)
```
Comment (mathurinm) [inline skglm/penalties/block_separable.py:186]: some parenthesis are not needed

File: skglm/penalties/block_separable.py [hunk 2]
```diff
@@ -153,3 +153,81 @@ def subdiff_distance(self, W, grad, ws):
...
+        value = sum_{j=1}^{n_features} pen(||W_j||)
+    """
+
+    def __init__(self, alpha, gamma):
+        self.alpha = alpha
+        self.gamma = gamma
+
+    def value(self, W):
+        """Compute the value of the SCAD penalty at W."""
+        n_features = W.shape[1]
```
Comment (mathurinm) [inline skglm/penalties/block_separable.py:184]: shape[0] no ?

File: skglm/penalties/separable.py [hunk 3]
```diff
@@ -174,7 +174,7 @@ class MCPenalty(BasePenalty):
     With x >= 0
     pen(x) =
     alpha * x - x^2 / (2 * gamma) if x =< gamma * alpha
-    gamma * alpha 2 / 2           if x > gamma * alpha
+    gamma * alpha ** 2 / 2           if x > gamma * alpha
```
Comment (mathurinm) [inline skglm/penalties/separable.py:177]: nice catch, just keep the ifs aligned

File: skglm/penalties/separable.py [hunk 4]
```diff
@@ -207,10 +207,9 @@ def subdiff_distance(self, w, grad, ws):
                 # distance of -grad to alpha * [-1, 1]
                 subdiff_dist[idx] = max(0, np.abs(grad[idx]) - self.alpha)
             elif np.abs(w[j]) < self.alpha * self.gamma:
-                # distance of -grad_j to (alpha - abs(w[j])/gamma) * sign(w[j])
+                # distance of -grad_j to (alpha * sign(w[j]) - w[j] / gamma)
```
Comment (mathurinm) [inline skglm/penalties/separable.py:210]: simpler?

File: skglm/penalties/separable.py [hunk 5]
```diff
@@ -229,6 +228,84 @@ def alpha_max(self, gradient0):
...
+    """Smoothly Clipped Absolute Deviation.
+
+    Notes
+    -----
+    With x >= 0
+    pen(x) =
+    alpha * x                         if x =< alpha
+    (2 * gamma * alpha * x - x ** 2 - alpha ** 2) \
+        / (2 * (gamma - 1))           if alpha < x < alpha * gamma
+    (alpha ** 2 * (gamma + 1)) / 2    if x > gamma * alpha
```
Comment (mathurinm) [inline skglm/penalties/separable.py:248]: rm parenthesis

File: skglm/penalties/separable.py [hunk 6]
```diff
@@ -229,6 +228,84 @@ def alpha_max(self, gradient0):
...
+    value = sum_{j=1}^{n_features} pen(abs(w_j))
+    """
+
+    def __init__(self, alpha, gamma):
+        self.alpha = alpha
+        self.gamma = gamma
+
+    def value(self, w):
+        """Compute the value of the SCAD penalty at w."""
+        value = np.full_like(w, (self.alpha ** 2 * (self.gamma + 1)) / 2)
```
Comment (mathurinm) [inline skglm/penalties/separable.py:258]: rm some parenthesis

File: skglm/penalties/separable.py [hunk 7]
```diff
@@ -229,6 +228,84 @@ def alpha_max(self, gradient0):
...
+        for j in range(len(w)):
+            if np.abs(w[j]) <= self.alpha:
+                value[j] = self.alpha * np.abs(w[j])
+            elif np.abs(w[j]) > self.alpha and np.abs(w[j]) < self.alpha * self.gamma:
```
Comment (mathurinm) [inline skglm/penalties/separable.py:262]: first condition can safely be ommitted due to being in an elif

File: skglm/penalties/separable.py [hunk 8]
```diff
@@ -229,6 +228,84 @@ def alpha_max(self, gradient0):
...
+                value[j] = (
+                    2 * self.gamma * self.alpha * np.abs(w[j])
+                    - np.abs(w[j]) ** 2 - self.alpha ** 2) / (2 * (self.gamma - 1))
```
Comment (mathurinm) [inline skglm/penalties/separable.py:265]: abs can be removed

File: skglm/penalties/separable.py [hunk 9]
```diff
@@ -229,6 +228,84 @@ def alpha_max(self, gradient0):
...
+                subdiff_dist[idx] = max(0, np.abs(grad[idx]) - self.alpha)
+            elif np.abs(w[j]) <= self.alpha:
+                # distance of -grad_j to alpha * sgn(w[j])
+                subdiff_dist[idx] = np.abs(grad[idx] + self.alpha * np.sign(w[j]))
+            elif np.abs(w[j]) > self.alpha and np.abs(w[j]) < self.alpha * self.gamma:
+                # distance of -grad_j to (alpha * gamma * sign(w[j]) - w[j])
+                #                        / (gamma - 1)
+                subdiff_dist[idx] = np.abs(
+                    grad[idx] +
+                    (self.alpha * self.gamma * np.sign(w[j]) - w[j]) / (self.gamma - 1)
```
Comment (mathurinm) [inline skglm/penalties/separable.py:293]: seems correct to me... I don't see where the problem comes from

File: skglm/penalties/separable.py [hunk 10]
```diff
@@ -229,6 +228,84 @@ def alpha_max(self, gradient0):
...
+
+    def prox_1d(self, value, stepsize, j):
+        """Compute the proximal operator of SCAD penalty."""
+        tau = self.alpha * stepsize
+        g = self.gamma / stepsize
+        if np.abs(value) <= 2 * tau:
+            return ST(value, tau)
+        if np.abs(value) > g * tau:
+            return value
+        return ((g - 1) / (g - 2)) * ST(value, (g * tau) / (g - 1))
```
Comment (mathurinm) [inline skglm/penalties/separable.py:276]: sure ? 
In wen's review I see 
image

(a is gamma) 
I'm not sure this matches what we have written

File: skglm/penalties/separable.py (hunk 9 above)
Comment (PABannier) [reply skglm/penalties/separable.py:293]: @Klopfe and I have triple checked the computations. now you. @QB3 ?

File: skglm/penalties/separable.py (hunk 10 above)
Comment (Klopfe) [reply skglm/penalties/separable.py]: calculation with stepsize is given here https://arxiv.org/pdf/1303.4434.pdf:
I do not think there is an easier way.

File: skglm/penalties/separable.py (hunk 8 above)
Comment (mathurinm) [reply skglm/penalties/separable.py:265]: There is a square: `np.abs(w[j]) ** 2 `

File: skglm/penalties/block_separable.py [hunk 11]
```diff
@@ -153,3 +153,91 @@ def subdiff_distance(self, W, grad, ws):
...
+            self.gamma - 1 - stepsize)
+        x_2 = np.abs(x_2)
+        x_3 = norm(value)
+
+        objs = np.zeros(3)
+        # TODO
+        objs[0] = 0.
+        objs[1] = 0.
+        objs[2] = 0.
+
```
Comment (PABannier) [inline skglm/penalties/block_separable.py:210]: @Klopfe i'm not too sure as to how translate the prox calculation to the multi-task case. `np.abs`translates to `norm`?

File: skglm/penalties/block_separable.py [hunk 12]
```diff
@@ -153,3 +153,91 @@ def subdiff_distance(self, W, grad, ws):
...
+        idx_min = np.argmin(objs)
+        if idx_min == 0:
+            return norm(value) * x_1
+        elif idx_min == 1:
+            return norm(value) * x_2
+        else:
+            return norm(value) * x_3
```
Comment (PABannier) [inline skglm/penalties/block_separable.py:217]: Here, not sure as well. `prox_1feat` should return a vector.

File: skglm/penalties/block_separable.py [hunk 13]
```diff
@@ -153,3 +153,91 @@ def subdiff_distance(self, W, grad, ws):
...
+                subdiff_dist[idx] = max(0, norm(grad[idx]) - self.alpha)
+            elif norm_Wj <= self.alpha:
+                # distance of -grad_j to alpha * W[j] / ||W[j]||
+                subdiff_dist[idx] = norm(grad[idx] + self.alpha * W[j] / norm_Wj)
+            elif norm_Wj <= self.gamma * self.alpha:
+                # distance of -grad_j to (alpha * gamma - ||W[j]||)
+                # / ((gamma - 1) * ||W[j]||) * W[j]
+                subdiff_dist[idx] = norm(grad[idx] + (
+                    (self.alpha * self.gamma - norm_Wj) / (norm_Wj * (self.gamma - 1))
+                ) * W[j])
```
Comment (PABannier) [inline skglm/penalties/block_separable.py:235]: @mathurinm @Klopfe could you double check this case?

File: toy.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import numpy as np
+
+from skglm.datafits import Quadratic
+from skglm.penalties import SCAD
+from skglm.solvers.cd_solver import cd_solver
+
+
+n_samples = 1000
+n_features = 100_000
+
```
Comment (PABannier) [inline toy.py:10]: After @Klopfe 's fix of prox computation, this toy script does converge in small and large settings, for arbitrarily small tolerance. If ok, I can remove it.

Comment (Klopfe) [reply toy.py:10]: ok for me

File: examples/plot_pen_prox.py [hunk 15]
```diff
@@ -32,7 +37,8 @@
                   label=pen.__class__.__name__)
 
 axarr[0].legend()
-axarr[1].legend()
+# axarr[1].legend()
 axarr[0].set_title("Penalty value")
 axarr[1].set_title("Proximal operator of penalty")
+fig.savefig("pen_prox.pdf")
```
Comment (mathurinm) [inline examples/plot_pen_prox.py:43]:
```suggestion
```

File: skglm/penalties/block_separable.py (hunk 11 above)
Comment (mathurinm) [reply skglm/penalties/block_separable.py:210]: check prop 18 in appendix D. Currently we're returning a scalar, it's not homogeneous
Does it mean the code is not tested or it manages to converge nevertheless ?

File: skglm/penalties/block_separable.py (hunk 12 above)
Comment (mathurinm) [reply skglm/penalties/block_separable.py:217]: OK yes, see my comment above

**Review by mathurinm — CHANGES_REQUESTED** [review]
One change in BlockSCAD, else LGTM

**Comment by PABannier** [comment]
update: blockSCAD does converge now, but very slowly. With `ws_strategy="fixpoint"`, it is faster. I suspect the distance to the subdifferential is wrong for one case (see the review above). 
I'd suspect there's still an error in the prox computation.

File: skglm/penalties/block_separable.py (hunk 13 above)
Comment (Klopfe) [reply skglm/penalties/block_separable.py:239]: Seems good to me!

**Review by Klopfe — APPROVED** [review]

**Comment by PABannier** [comment]
From the experiments I performed, no. Now it works fine. I'd be glad if you could tell me if it works fine for you.

**Comment by Klopfe** [comment]
> From the experiments I performed, no. Now it works fine. I'd be glad if you could tell me if it works fine for you as well.

I'll run some tests and let you know. It would be nice to merge this PR soon. Thanks for the great work!

**Comment by mathurinm** [comment]
@PABannier can you add unit tests for Scad and BlockSCAD ?

**Comment by QB3** [comment]
> @PABannier can you add unit tests for Scad and BlockSCAD ?

Do we know how we could test SCAD @PABannier @mathurinm ? It is not so clear to me.

File: skglm/tests/test_penalties.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    IndicatorBox(alpha=alpha),
+    L0_5(alpha),
+    L2_3(alpha)]
+
+
+@pytest.mark.parametrize('penalty', penalties)
+def test_subdiff_diff(penalty):
+    estimator_ours = GeneralizedLinearEstimator(
+        datafit=Quadratic(),
+        penalty=L1(alpha=1),
```
[the remark this answers is not in the document]
Comment (QB3) [reply skglm/tests/test_penalties.py]: yes! this is fixed in the next commit.

File: skglm/penalties/block_separable.py [hunk 17]
```diff
@@ -156,6 +156,98 @@ def is_penalized(self, n_features):
...
+                value[j] = self.alpha * norm_rows[j]
+            elif norm_rows[j] <= self.alpha * self.gamma:
+                value[j] = (
+                    2 * self.gamma * self.alpha * norm_rows[j] - norm_rows[j] ** 2
+                    - self.alpha ** 2) / (2 * (self.gamma - 1))
+        return np.sum(value)
+
+    def prox_1feat(self, value, stepsize, j):
+        """Compute the proximal operator of BlockSCAD."""
+        nrm_value = norm(value)
```
Comment (mathurinm) [inline skglm/penalties/block_separable.py:199]: The formula here is the 1D SCAD prox applied to norm(value), times value / norm(value)

So we can factorize code bu putting a function prox_scad in utils, and use it both here and in single task scad.

File: skglm/tests/test_penalties.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+block_penalties = [
+    L2_1(alpha=alpha), L2_05(alpha=alpha),
+    BlockMCPenalty(alpha=alpha, gamma=4),
+    BlockSCAD(alpha=alpha, gamma=4)
+    ]
+
+
+@pytest.mark.parametrize('penalty', penalties)
+def test_subdiff_diff(penalty):
+    estimator_ours = GeneralizedLinearEstimator(
```
Comment (mathurinm) [inline skglm/tests/test_penalties.py:46]: call it clf for simplicity ? as you wish

Comment (QB3) [reply skglm/tests/test_penalties.py]: I went for est (for me clf stands for classifier)

Comment (mathurinm) [reply skglm/tests/test_penalties.py]: perfect

File: skglm/penalties/block_separable.py (hunk 17 above)
Comment (QB3) [reply skglm/penalties/block_separable.py]: I have done it for SCAD and the MCP

File: skglm/tests/test_penalties.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        datafit=QuadraticMultiTask(),
+        penalty=block_penalty,
+        tol=1e-14,
+    ).fit(X, Y)
+    # assert the stopping criterion is satisfied
+    assert_array_less(est.stop_crit_, est.tol)
+
+
+if __name__ == '__main__':
+    # test_subdiff_diff(L1(alpha=alpha))
```
Comment (mathurinm) [inline skglm/tests/test_penalties.py:67]: cleanup

File: skglm/utils.py [hunk 20]
```diff
@@ -43,6 +43,69 @@ def box_proj(x, low, up):
...
+    """Compute the value of MCP."""
+    s0 = np.abs(w) < gamma * alpha
+    value = np.full_like(w, gamma * alpha ** 2 / 2.)
+    value[s0] = alpha * np.abs(w[s0]) - w[s0]**2 / (2 * gamma)
+    return np.sum(value)
+
+
+@njit
+def prox_MCP(value, stepsize, alpha, gamma):
+    """Compute the proximal operator of MCP penalty."""
```
Comment (mathurinm) [inline skglm/utils.py:57]: to be clear: of stepsize * MCP ?

File: skglm/utils.py [hunk 21]
```diff
@@ -43,6 +43,69 @@ def box_proj(x, low, up):
...
+        elif np.abs(w[j]) <= alpha * gamma:
+            value[j] = (
+                2 * gamma * alpha * np.abs(w[j])
+                - w[j] ** 2 - alpha ** 2) / (2 * (gamma - 1))
+    return np.sum(value)
+
+
+@njit
+def prox_SCAD(value, stepsize, alpha, gamma):
+    """Compute the proximal operator of SCAD penalty."""
```
Comment (mathurinm) [inline skglm/utils.py:83]: of stepsize * SCAD ?

File: skglm/utils.py [hunk 22]
```diff
@@ -43,6 +43,69 @@ def box_proj(x, low, up):
...
+    tau = gamma * alpha
+    x_1 = max(0, np.abs(value) - alpha * stepsize)
+    x_2 = ((gamma - 1) * np.abs(value) - stepsize * tau) / (
+        gamma - 1 - stepsize)
+    x_2 = abs(x_2)
+
+    x_3 = abs(value)
+
+    objs = np.zeros(3)
```
Comment (mathurinm) [inline skglm/utils.py:92]: can this be made simpler with a list comprehension (for length 3) ?

File: skglm/utils.py [hunk 23]
```diff
@@ -43,6 +43,69 @@ def box_proj(x, low, up):
...
+    objs[0] = (0.5 / stepsize) * (
+        x_1 - np.abs(value)) ** 2 + value_SCAD(np.array([x_1]), alpha, gamma)
+    objs[1] = (0.5 / stepsize) * (
+        x_2 - np.abs(value)) ** 2 + value_SCAD(np.array([x_2]), alpha, gamma)
+    objs[2] = (0.5 / stepsize) * (
+        x_3 - np.abs(value)) ** 2 + value_SCAD(np.array([x_3]), alpha, gamma)
+
+    idx_min = np.argmin(objs)
+    if idx_min == 0:
```
Comment (mathurinm) [inline skglm/utils.py:101]: can this be made simpler by first putting x1, x2 and x3 into a list ? then x_list[np.argmin(values])]

File: skglm/utils.py [hunk 24]
```diff
@@ -43,6 +43,57 @@ def box_proj(x, low, up):
         return x
 
 
+@njit
+def value_MCP(w, alpha, gamma):
```
Comment (Badr-MOUFAD) [inline skglm/utils.py:47]: Is there a reason behind defining value and prox functions in a separate file (``utils.py``) and not within the class definition?

Comment (mathurinm) [reply skglm/utils.py:47]: yes, they are used both in single task and group penalties

**Comment by Klopfe** [comment]
> I wish we could factoriz the subdiff and block subdiff, but it does not seem trivial to me. Apart from this, LGTM.

yes, I spent quite a while trying to do so but it does not seem so easy... I think we can leave it as it is right now.

**Review by mathurinm — APPROVED** [review]

**Pull request merged**